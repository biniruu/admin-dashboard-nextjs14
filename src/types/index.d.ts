import { type ReactNode } from 'react'

export interface CardData {
  icon: ReactNode
  title: string
  number: number
  state: 'positive' | 'negative'
  rate: number
  detail: string
}

export interface Product {
  id?: string
  img?: string
  size?: string
  color?: string
  title: string
  desc: string
  price: string
  stock: string
  createdAt?: Date
  updatedAt?: Date
}

export interface User {
  id?: string
  img?: string
  username: string
  email: string
  phone: string
  address: string
  password: string
  isAdmin: boolean
  isActive: boolean
  createdAt?: Date
  updatedAt?: Date
}
