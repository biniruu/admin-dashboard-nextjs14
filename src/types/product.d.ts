export interface Product {
  id: string
  img: string
  title: string
  desc: string
  price: string
  createdAt: number
  stock: string
}
