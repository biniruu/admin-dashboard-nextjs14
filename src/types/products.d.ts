export interface Products {
  id: string
  img: string
  title: string
  desc: string
  price: string
  createdAt: number
  stock: string
}
