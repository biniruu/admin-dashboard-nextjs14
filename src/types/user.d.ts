export interface User {
  img: string
  username: string
  id: string
  email: string
  phone: string
  address: string
  isAdmin: boolean
  isActive: boolean
}
