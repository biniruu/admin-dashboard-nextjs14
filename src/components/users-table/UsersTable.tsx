'use client'

import dayjs from 'dayjs'
import Image from 'next/image'
import Link from 'next/link'

import styles from './usersTable.module.css'

import { type User } from 'types'

const tableHead: string[] = ['name', 'email', 'created at', 'role', 'status', 'action']

interface Props {
  users: User[]
  deleteUser: (formData: FormData) => Promise<void>
}

function UsersTable({ users, deleteUser }: Props) {
  return (
    <table className={`w-full ${styles.table}`}>
      <thead>
        <tr>
          {tableHead.map(item => (
            <td key={item}>{item}</td>
          ))}
        </tr>
      </thead>
      <tbody>
        {users?.map(user => {
          const { _id: id, img, username, email, isAdmin, isActive, createdAt } = user
          const date = createdAt && dayjs(createdAt).format('YYYY/MM/DD')

          return (
            <tr key={id}>
              <td>
                <div className="flex items-center gap-default">
                  <Image
                    src={img || '/noavatar.png'}
                    alt=""
                    width={40}
                    height={40}
                    className="rounded-half object-cover"
                  />
                  {username}
                </div>
              </td>
              <td>{email}</td>
              <td>{date || ''}</td>
              <td>{isAdmin ? 'admin' : 'user'}</td>
              <td>{isActive ? 'active' : 'passive'}</td>
              <td className="flex gap-default">
                <Link href="/">
                  <button className={`${styles.button} bg-[teal]`}>view</button>
                </Link>
                <form action={deleteUser}>
                  <input type="hidden" name="id" value={id} />
                  <button className={`${styles.button} bg-[crimson]`}>delete</button>
                </form>
              </td>
            </tr>
          )
        })}
      </tbody>
    </table>
  )
}

export default UsersTable
