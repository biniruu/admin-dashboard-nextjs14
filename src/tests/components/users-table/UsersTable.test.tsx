import renderer from 'react-test-renderer'

import UsersTable from '@components/users-table/UsersTable'
import { User } from 'types'

jest.mock('lib/actions', () => ({
  deleteUser: jest.fn(),
}))

describe('UsersTable component', () => {
  let users: User[]

  beforeEach(() => {
    // It doesn't need to use some properties: password, updatedAt, phone, address. However it is used for just 'User' type
    users = [
      {
        id: 'id1',
        img: '',
        username: 'username1',
        email: 'email1@email.com',
        isAdmin: true,
        isActive: true,
        createdAt: new Date(),
        password: '',
        updatedAt: new Date(),
        phone: '',
        address: '',
      },
      {
        id: 'id2',
        img: '',
        username: 'username2',
        email: 'email2@email.com',
        isAdmin: false,
        isActive: false,
        createdAt: new Date(),
        password: '',
        updatedAt: new Date(),
        phone: '',
        address: '',
      },
    ]
  })

  test('should render the component', () => {
    const result = renderer.create(<UsersTable users={users} />).toJSON()

    expect(result).toMatchSnapshot()
  })
})
