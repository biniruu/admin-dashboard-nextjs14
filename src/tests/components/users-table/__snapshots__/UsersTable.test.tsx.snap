// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersTable component should render the component 1`] = `
<table
  className="w-full table"
>
  <thead>
    <tr>
      <td>
        name
      </td>
      <td>
        email
      </td>
      <td>
        created at
      </td>
      <td>
        role
      </td>
      <td>
        status
      </td>
      <td>
        action
      </td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <div
          className="flex items-center gap-default"
        >
          <img
            alt=""
            className="rounded-half object-cover"
            data-nimg="1"
            decoding="async"
            height={40}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="/_next/image?url=%2Fnoavatar.png&w=96&q=75"
            srcSet="/_next/image?url=%2Fnoavatar.png&w=48&q=75 1x, /_next/image?url=%2Fnoavatar.png&w=96&q=75 2x"
            style={
              {
                "color": "transparent",
              }
            }
            width={40}
          />
          username1
        </div>
      </td>
      <td
        className="normal-case"
      >
        email1@email.com
      </td>
      <td>
        2024/04/24
      </td>
      <td>
        admin
      </td>
      <td>
        active
      </td>
      <td>
        <div
          className="flex gap-default"
        >
          <a
            href="/dashboard/users/id1"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            <button
              className="button bg-[teal]"
            >
              view
            </button>
          </a>
          <form
            action={[MockFunction]}
          >
            <input
              name="id"
              type="hidden"
              value="id1"
            />
            <button
              className="button bg-[crimson]"
            >
              delete
            </button>
          </form>
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div
          className="flex items-center gap-default"
        >
          <img
            alt=""
            className="rounded-half object-cover"
            data-nimg="1"
            decoding="async"
            height={40}
            loading="lazy"
            onError={[Function]}
            onLoad={[Function]}
            src="/_next/image?url=%2Fnoavatar.png&w=96&q=75"
            srcSet="/_next/image?url=%2Fnoavatar.png&w=48&q=75 1x, /_next/image?url=%2Fnoavatar.png&w=96&q=75 2x"
            style={
              {
                "color": "transparent",
              }
            }
            width={40}
          />
          username2
        </div>
      </td>
      <td
        className="normal-case"
      >
        email2@email.com
      </td>
      <td>
        2024/04/24
      </td>
      <td>
        user
      </td>
      <td>
        passive
      </td>
      <td>
        <div
          className="flex gap-default"
        >
          <a
            href="/dashboard/users/id2"
            onClick={[Function]}
            onMouseEnter={[Function]}
            onTouchStart={[Function]}
          >
            <button
              className="button bg-[teal]"
            >
              view
            </button>
          </a>
          <form
            action={[MockFunction]}
          >
            <input
              name="id"
              type="hidden"
              value="id2"
            />
            <button
              className="button bg-[crimson]"
            >
              delete
            </button>
          </form>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
